haystack/components/validators/json_schema.py

```diff
@@ -0,0 +1,172 @@
+
+    If JSON content of a message conforms to the provided schema, the message is passed along the "validated" output.
+    If the JSON content does not conform to the schema, the message is passed along the "validation_error" output.
+    In the latter case, the error message is constructed using the provided error_template or a default template.
+    These error ChatMessages can be used by LLMs in Haystack 2.x recovery loops.
+    """
+
+    # Default error description template
+    default_error_template = (
+        "The JSON content in the previous message does not conform to the provided schema.\n"
+        "Error details:\n- Message: {error_message}\n"
+        "- Error Path in JSON: {error_path}\n"
+        "- Schema Path: {error_schema_path}\n"
+        "Please match the following schema:\n"
+        "{json_schema}\n"
+        "and provide the corrected JSON content ONLY."
+    )
+
+    def __init__(self):
+        jsonschema_import.check()
+
+    @component.output_types(validated=List[ChatMessage], validation_error=List[ChatMessage])
+    def run(
+        self,
+        messages: List[ChatMessage],
+        json_schema: Dict[str, Any],
+        previous_messages: Optional[List[ChatMessage]] = None,
+        error_template: Optional[str] = None,
+    ):
+        """
+        Checks if the last message and its content field conforms to json_schema.
+
+        :param messages: A list of ChatMessage instances to be validated. The last message in this list is the one
+        that is validated.
+        :param previous_messages: A list of previous ChatMessage instances, by default None. These are not validated
+        but are returned in the case of an error.
+        :param json_schema:A dictionary representing the JSON schema against which the messages' content is validated.
+        :param error_template: A custom template string for formatting the error message in case of validation
+        failure, by default None.
+        """
+        last_message = messages[-1]
+        last_message_content = json.loads(last_message.content)
+
+        # fc payload is json object but subtree `parameters` is string - we need to convert to json object
+        # we need complete json to validate it against schema
+        last_message_json = self.recursive_json_to_object(last_message_content)
+        using_openai_schema: bool = self.is_openai_function_calling_schema(json_schema)
+        if self.is_openai_function_calling_schema(json_schema):
+            validation_schema = json_schema["parameters"]
+        else:
+            validation_schema = json_schema
+        try:
+            last_message_json = [last_message_json] if not isinstance(last_message_json, list) else last_message_json
+            for content in last_message_json:
+                if not self.is_function_calling_payload(content):
+                    raise ValidationError(f"{content} is not a valid OpenAI function calling payload.")
```

Comment:
This makes this component specific for function calling.
Is it wanted? Or is it an error?

Reply:
Great catch; why constrict it only to function calling? Was biased by my particular use case. See pydocs class example for generic json validation